commit 17eae09f1e8c5fcef343923cd8c0e98f1028849e
Author: Walter Harley <wharley>
Date:   Sat Mar 29 00:15:04 2008 +0000

    Bug 224750 - NPE while annotation processing in reconcile, if annotation type is not yet imported

2	2	org.eclipse.jdt.apt.core/src/org/eclipse/jdt/apt/core/internal/declaration/AnnotationMirrorImpl.java
2	2	org.eclipse.jdt.apt.core/src/org/eclipse/jdt/apt/core/internal/declaration/ClassDeclarationImpl.java
2	2	org.eclipse.jdt.apt.core/src/org/eclipse/jdt/apt/core/internal/declaration/ExecutableUtil.java
2	2	org.eclipse.jdt.apt.core/src/org/eclipse/jdt/apt/core/internal/declaration/TypeDeclarationImpl.java
1	1	org.eclipse.jdt.apt.core/src/org/eclipse/jdt/apt/core/internal/env/AbstractCompilationEnv.java
2	2	org.eclipse.jdt.apt.core/src/org/eclipse/jdt/apt/core/internal/env/BuildEnv.java
diff --git a/org.eclipse.jdt.apt.core/src/org/eclipse/jdt/apt/core/internal/declaration/AnnotationMirrorImpl.java b/org.eclipse.jdt.apt.core/src/org/eclipse/jdt/apt/core/internal/declaration/AnnotationMirrorImpl.java
index ee6fcae..0f71084 100644
--- a/org.eclipse.jdt.apt.core/src/org/eclipse/jdt/apt/core/internal/declaration/AnnotationMirrorImpl.java
+++ b/org.eclipse.jdt.apt.core/src/org/eclipse/jdt/apt/core/internal/declaration/AnnotationMirrorImpl.java
@@ -1,5 +1,5 @@
 /*******************************************************************************
- * Copyright (c) 2005, 2007 BEA Systems, Inc.
+ * Copyright (c) 2005, 2008 BEA Systems, Inc.
  * All rights reserved. This program and the accompanying materials
  * are made available under the terms of the Eclipse Public License v1.0
  * which accompanies this distribution, and is available at
@@ -63,7 +63,7 @@ public class AnnotationMirrorImpl implements AnnotationMirror, EclipseMirrorObje
     public AnnotationType getAnnotationType()
     {		
         final ITypeBinding binding = _domAnnotation.getAnnotationType();
-        if( binding == null ){
+        if( binding == null || binding.isRecovered() ){
         	final ASTNode node = _annotated.getCompilationUnit().findDeclaringNode(_domAnnotation);
         	String name = ""; //$NON-NLS-1$
         	if( node != null && node instanceof Annotation ){
diff --git a/org.eclipse.jdt.apt.core/src/org/eclipse/jdt/apt/core/internal/declaration/ClassDeclarationImpl.java b/org.eclipse.jdt.apt.core/src/org/eclipse/jdt/apt/core/internal/declaration/ClassDeclarationImpl.java
index 02e9c0e..4b97436 100644
--- a/org.eclipse.jdt.apt.core/src/org/eclipse/jdt/apt/core/internal/declaration/ClassDeclarationImpl.java
+++ b/org.eclipse.jdt.apt.core/src/org/eclipse/jdt/apt/core/internal/declaration/ClassDeclarationImpl.java
@@ -1,5 +1,5 @@
 /*******************************************************************************
- * Copyright (c) 2005, 2007 BEA Systems, Inc.
+ * Copyright (c) 2005, 2008 BEA Systems, Inc.
  * All rights reserved. This program and the accompanying materials
  * are made available under the terms of the Eclipse Public License v1.0
  * which accompanies this distribution, and is available at
@@ -135,7 +135,7 @@ public class ClassDeclarationImpl extends TypeDeclarationImpl implements ClassDe
         final ITypeBinding superClass = getDeclarationBinding().getSuperclass();
 		if ( superClass == null )
 			return null;
-		else if( superClass.isClass() )
+		else if( superClass.isClass() && !superClass.isRecovered() )
             return (ClassType)Factory.createReferenceType(superClass, _env);
         else // catch error case where user extends some interface instead of a class.
             return Factory.createErrorClassType(superClass);
diff --git a/org.eclipse.jdt.apt.core/src/org/eclipse/jdt/apt/core/internal/declaration/ExecutableUtil.java b/org.eclipse.jdt.apt.core/src/org/eclipse/jdt/apt/core/internal/declaration/ExecutableUtil.java
index b894ce1..4f5a044 100644
--- a/org.eclipse.jdt.apt.core/src/org/eclipse/jdt/apt/core/internal/declaration/ExecutableUtil.java
+++ b/org.eclipse.jdt.apt.core/src/org/eclipse/jdt/apt/core/internal/declaration/ExecutableUtil.java
@@ -1,5 +1,5 @@
 /*******************************************************************************
- * Copyright (c) 2005, 2007 BEA Systems, Inc.
+ * Copyright (c) 2005, 2008 BEA Systems, Inc.
  * All rights reserved. This program and the accompanying materials
  * are made available under the terms of the Eclipse Public License v1.0
  * which accompanies this distribution, and is available at
@@ -185,7 +185,7 @@ class ExecutableUtil {
 	    	for(Name exception : exceptions ){
 	    		final ITypeBinding eType = exception.resolveTypeBinding();
 	    		final ReferenceType refType;
-	    		if( eType == null )
+	    		if( eType == null || eType.isRecovered() )
 	    			refType = Factory.createErrorClassType(exception.toString());
 	    		else
 	    			refType = Factory.createReferenceType(eType, env);
diff --git a/org.eclipse.jdt.apt.core/src/org/eclipse/jdt/apt/core/internal/declaration/TypeDeclarationImpl.java b/org.eclipse.jdt.apt.core/src/org/eclipse/jdt/apt/core/internal/declaration/TypeDeclarationImpl.java
index 8bf4e35..071a0d9 100644
--- a/org.eclipse.jdt.apt.core/src/org/eclipse/jdt/apt/core/internal/declaration/TypeDeclarationImpl.java
+++ b/org.eclipse.jdt.apt.core/src/org/eclipse/jdt/apt/core/internal/declaration/TypeDeclarationImpl.java
@@ -1,5 +1,5 @@
 /*******************************************************************************
- * Copyright (c) 2005, 2007 BEA Systems, Inc.
+ * Copyright (c) 2005, 2008 BEA Systems, Inc.
  * All rights reserved. This program and the accompanying materials
  * are made available under the terms of the Eclipse Public License v1.0
  * which accompanies this distribution, and is available at
@@ -216,7 +216,7 @@ public abstract class TypeDeclarationImpl extends MemberDeclarationImpl
         for( ITypeBinding binding : superInterfaceBindings ){
             if( binding.isInterface() ){
                 final TypeDeclarationImpl mirrorDecl = Factory.createReferenceType(binding, _env);
-                if( mirrorDecl.kind() == MirrorKind.TYPE_INTERFACE ){
+                if( mirrorDecl != null && mirrorDecl.kind() == MirrorKind.TYPE_INTERFACE ){
                     results.add((InterfaceType)mirrorDecl);
                 }
             }
diff --git a/org.eclipse.jdt.apt.core/src/org/eclipse/jdt/apt/core/internal/env/AbstractCompilationEnv.java b/org.eclipse.jdt.apt.core/src/org/eclipse/jdt/apt/core/internal/env/AbstractCompilationEnv.java
index b1bc62f..ded0518 100644
--- a/org.eclipse.jdt.apt.core/src/org/eclipse/jdt/apt/core/internal/env/AbstractCompilationEnv.java
+++ b/org.eclipse.jdt.apt.core/src/org/eclipse/jdt/apt/core/internal/env/AbstractCompilationEnv.java
@@ -272,7 +272,7 @@ public abstract class AbstractCompilationEnv
 				continue;
 			final TypeDeclarationImpl decl = 
 				Factory.createReferenceType(annoType, this);
-			if (decl.kind() == EclipseMirrorObject.MirrorKind.TYPE_ANNOTATION){
+			if (decl != null && decl.kind() == EclipseMirrorObject.MirrorKind.TYPE_ANNOTATION){
 				final AnnotationTypeDeclaration annoDecl = (AnnotationTypeDeclaration)decl;
 				decls.put(annoDecl.getQualifiedName(), annoDecl);
 			}
diff --git a/org.eclipse.jdt.apt.core/src/org/eclipse/jdt/apt/core/internal/env/BuildEnv.java b/org.eclipse.jdt.apt.core/src/org/eclipse/jdt/apt/core/internal/env/BuildEnv.java
index 635400e..409fea8 100644
--- a/org.eclipse.jdt.apt.core/src/org/eclipse/jdt/apt/core/internal/env/BuildEnv.java
+++ b/org.eclipse.jdt.apt.core/src/org/eclipse/jdt/apt/core/internal/env/BuildEnv.java
@@ -1,5 +1,5 @@
 /*******************************************************************************
- * Copyright (c) 2005, 2007 BEA Systems, Inc.
+ * Copyright (c) 2005, 2008 BEA Systems, Inc.
  * All rights reserved. This program and the accompanying materials
  * are made available under the terms of the Eclipse Public License v1.0
  * which accompanies this distribution, and is available at
@@ -304,7 +304,7 @@ public class BuildEnv extends AbstractCompilationEnv
 					continue;
 				final TypeDeclarationImpl decl = 
 					Factory.createReferenceType(annoType, this);
-				if (decl.kind() == EclipseMirrorObject.MirrorKind.TYPE_ANNOTATION){
+				if (decl != null && decl.kind() == EclipseMirrorObject.MirrorKind.TYPE_ANNOTATION){
 					final AnnotationTypeDeclaration annoDecl = (AnnotationTypeDeclaration)decl;
 					decls.put(annoDecl.getQualifiedName(), annoDecl);
 					perFileAnnos.add(annoDecl);
